add support for write index resolution when creating/updating documents

talevy opened the pull request:
This commit introduces a new option to IndicesOptions that requires
a write index to exist and then uses this option for index/update requests.
Get requests ignore the write index.

elasticmachine commented:
Pinging @elastic/es-core-infra

bleskes commented:
@talevy thx. I had a look and some thinking time and I would like to take a somewhat (but not complete) different approach.

1) Can you please extract the bug fix to it's own PR? this way we can raise awareness, look it up etc.
2) I think indices option as it is today has a problem - it encapsulates both what the user wants (resolve wild cards, don't allow empty lists etc.) with what the API supports (can or can not work on closed indices, supports multiple indices). I think the later should be explicit and hard coded on the transport action level and not a per request thing. Since this is an existing issue, It's beyond the scope of this change to go and tackle that, but I don't want to make worse either by adding an option to `IndicesOption`. Instead, I propose adding a parameter to the  internal `concreteIndices` method of `IndexNameExpressionResolver` that controls whether aliases should use just the write index or all of the indices. Then we can add a public method ala `concreteSingleIndex` (call it `concreteWriteIndex`?) that uses that flag. 

How does that sound?

talevy commented:
@bleskes thanks for taking a look

1. will do; it was just blocking tests, so I decided to lump it together
2. I agree, it felt weird for sure, I just thought it would be more confusing to have two places "defining" the indices that are to be resolved to, and IndicesOptions felt like the right place for this reason. It would be nice to be able to leave a comment on IndicesOptions to explain this viewpoint so that it is clear that it is not to be the end-all definition for what is allowed. On its own, it is not clear. I'll update

talevy commented:
Another assumption I had was that `IndicesOptions.fromRequest` was being used by the index apis. Looking at it now, I was wrong, so mixing indicesOptions is irrelevant since users cannot change them anyways.

talevy commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexSourceTargetValidationTests.java:
```diff
@@ -46,13 +46,15 @@
  */
 public class ReindexSourceTargetValidationTests extends ESTestCase {
```
This class needs more work, but I am not sure whether we want reindex target-aliases to resolve to 
the write index. It feels right, but it also sounds scary. what do you think?

talevy replied:
oh nevermind, it *must* check details since even a single index with `is_write_index=false` should not be chosen by reindex api

talevy commented:
@bleskes I've updated to remove the flag from IndicesOptions. I still have some follow-up work to do with handling ReindexRequests, but I think it is in a place that can be reviewed.

Specifically regarding these two points

1. concreteIndices resolution logic
2. whether concreteIndexNames should be updated to include the writeIndex flag. I still haven't had CI run through all of the integration tests, so I may have to do this whether I would like to or not, depending on whether delete/update/index request tests leverage this

javanna commented:
I haven't looked into this in depth, but I see changes being made in how some actions resolve indices which makes me think that these changes should also be taken into account in the security plugin indices resolution code (see https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java). Have these changes been tested with security installed?

The reason why `concreteSingleIndex` as a separate method works ok with security is that it doesn't have any additional options compared to the standard `concreteIndices`. But in security we rely on the fact that every API resolves indices in the same way based on the provided indices options.

As a long-term plan it would be beneficial to unify indices resolution as described in #29915, so that this type of changes are automatically reflected in the security indices resolution logic (or hopefully security gets to have less and less of its own logic). I never got to work on that unfortunately, not even sure how hard that's going to be.

I would love if we could also work on #9438 as a follow-up, I am happy that we did not end up adding yet another option to `IndicesOptions` (though we have yet to see what needs to be done to adapt the security indices resolver), and I think the most important part of #9438 would be to separate the options that can be set by users from the ones that are internal and hardcoded for each API (along the lines of what Boaz mentioned above).

talevy commented:
thanks for providing all these details @javanna! I wasn't aware of any of that. I'll double check how Security interacts with this resolution and be sure it is tested

javanna commented:
Just chatted to @bleskes and we agreed that this deserves a test, but we think that this change should not impact security as it only has to do with how aliases get expanded, while our security plugin leaves aliases untouched.

bleskes commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -295,7 +295,7 @@ protected void doRun() throws Exception {
                             break;
                         case DELETE:
-                            docWriteRequest.routing(metaData.resolveIndexRouting(docWriteRequest.routing(), docWriteRequest.index()));
+                            docWriteRequest.routing(metaData.resolveIndexRouting(docWriteRequest.routing(), docWriteRequest.index(), true));
```
With some more thinking, I don't think we should make this feature (routing value in an alias) work with the write index flag. The problem is that if we allow to write to multi index aliases with this feature turned on, there is no way for get to work through those aliases without us changing the get behavior of resolving multi-index aliased base on the index write flag which is unexpected. Bottom line, can we see what it takes to make sure resolveIndexRouting explodes as soon as you touch more than one index via an alias?

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -117,7 +117,26 @@ public IndexNameExpressionResolver(Settings settings) {
+     * @throws IllegalArgumentException if one of the aliases resolve to multiple indices and the provided
+     * indices options in the context don't allow such a case.
+     */
+    public Index[] concreteIndices(ClusterState state, IndicesOptions options, boolean resolveToWriteIndex, String... indexExpressions) {
```
why does this method need to be public ? can we remove it and inline the implementation with it's single caller? it's just two lines.

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -194,29 +213,43 @@ public IndexNameExpressionResolver(Settings settings) {
             }
 
             Collection<IndexMetaData> resolvedIndices = aliasOrIndex.getIndices();
```
can we make if clause have this structure and inline this variable in the second else (see marker):

```
 if (aliasOrIndex.isAlias() && context.isResolveToWriteIndex()) {
    .. write operation path
} else {
  ... read path ..
  start the old code here and put the resolvedIndices variable definition here.
  
}
```

bleskes replied:
ping about inlining this variable.

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -255,6 +288,24 @@ public Index concreteSingleIndex(ClusterState state, IndicesRequest request) {
+        String indexExpression = request.indices() != null && request.indices().length > 0 ? request.indices()[0] : null;
+        Index[] indices = concreteIndices(state, request.indicesOptions(), true, indexExpression);
+        if (indices.length != 1) {
+            throw new IllegalArgumentException("unable to return a single index as the index and options provided got resolved to multiple indices");
```
I think we need to adapt this message for more things that can go wrong (single  index with no write flag etc.)

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -477,7 +477,7 @@ private static String mergePaths(String path, String field) {
     // TODO: This can be moved to IndexNameExpressionResolver too, but this means that we will support wildcards and other expressions
     // in the index,bulk,update and delete apis.
-    public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex) {
+    public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex, boolean isWriteOperation) {
```
see my comment earlier

talevy commented:
thanks for the reviews and follow-up @bleskes & @javanna. I am prioritizing some other tasks at the moment, but hope to get back with a follow-up to this tomorrow!

talevy commented:
re Reindex being orthogonal, that may be the case. I will double check I am understanding the logic.

This code seemed relevant for resolving the concrete target index: https://github.com/elastic/elasticsearch/blob/master/modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportReindexAction.java#L185

talevy commented:
@javanna maybe you know more, but I was speaking to @jaymode about the security aspects of resolving aliases. From our discussion, it did not seem like this was an issue since the security resolution never expanded aliases and passes down the alias to the regular resolution facilities.

One place that we did think would be reasonable to test was that `authorize` is successful on index/delete requests on aliases. Given the first paragraph above, I do not see the value in this.

Do you have a specific type of test that you have in mind?

something like this

```diff
@@ -1010,6 +1011,39 @@ public class AuthorizationServiceTests extends ESTestCase {

+    public void testAuthorizationOfIndividualAliasResolvedToWriteIndex() {
+        final String action = BulkAction.NAME + "[s]";
+        final BulkItemRequest[] items = {
+            new BulkItemRequest(1, new IndexRequest("concrete-index", "doc", "c2")),
+            new BulkItemRequest(2, new IndexRequest("alias-1", "doc", "a2b"))
+        };
+        final ShardId shardId = new ShardId("concrete-index", UUID.randomUUID().toString(), 1);
+        final TransportRequest request = new BulkShardRequest(shardId, WriteRequest.RefreshPolicy.IMMEDIATE, items);
+
+        final Authentication authentication = createAuthentication(new User("user", "my-role"));
+        RoleDescriptor role = new RoleDescriptor("my-role", null, new IndicesPrivileges[] {
+            IndicesPrivileges.builder().indices("concrete-index").privileges("all").build(),
+            IndicesPrivileges.builder().indices("alias-1").privileges("index").build()
+        }, null);
+        roleMap.put("my-role", role);
+
+        MetaData metaData = MetaData.builder()
+            .put(IndexMetaData.builder("concrete-index")
+            .settings(settings(Version.CURRENT)).numberOfReplicas(0).numberOfShards(1)
+            .putAlias(AliasMetaData.builder("alias-1").writeIndex(randomBoolean() ? null : randomBoolean())))
+            .build();
+
+        ClusterState state = mock(ClusterState.class);
+        when(clusterService.state()).thenReturn(state);
+        when(state.metaData()).thenReturn(metaData);
+
+        authorize(authentication, action, request);
+
+        //verify(auditTrail).accessGranted(authentication, IndexAction.NAME, request, new String[] { role.getName() }); // alias-1 index
+        verify(auditTrail).accessGranted(authentication, action, request, new String[] { role.getName() }); // bulk request is allowed
+        verifyNoMoreInteractions(auditTrail);
+    }
+
```

talevy commented:
@bleskes, regarding the index-routing-resolution, I decided that it was not a good place to be the judge of whether the alias had to point to a single index... Instead, I relaxed it to rejecting the case where it points to multiple indices, and any one of those indices has routing specified in their alias metadata. Since there are checks outside of this for single-index operation restraints, I feel like this is an appropriate relaxation here. what do you think?

javanna commented:
@talevy see my [comment](https://github.com/elastic/elasticsearch/pull/31520#issuecomment-400604343) above, I agree that we should be ok given how the security plugin works. I think that an integration test is what guarantees that things work well with security installed. Though given that we are modifying existing APIs that should already have their yaml tests (which we also run with security installed), I think a green build is a good indication that everything is fine. Thoughts @jaymode ?

talevy commented:
ah ok, @javanna. I wasn't sure because you also said this deserves a test. What I will do is add an extra set of yaml tests for bulk with is_write_index. I thought these were covered in IT tests, but security does not run those

javanna commented:
@talevy part of the confusion is because I haven't followed this very closely so I am unsure of what is already covered by existing tests and what tests should be added. But luckily security is good as-is, hopefully :)

jaymode commented:
I am good with having tests for this but I think we may need specific security tests for writes to an alias that the user doesn't have access to, writes when the user only has access to the alias and not the underlying index, etc.

talevy commented:
thanks @jaymode 

> I am good with having tests for this but I think we may need specific security tests for writes to an alias that the user doesn't have access to, writes when the user only has access to the alias and not the underlying index, etc.

this feels like a general test that is independent of these changes and out of scope for this PR, or do you think this should be added here?

talevy commented:
ack. then I will work on adding yaml tests that will check alias resolution + specific permission scenarios.

talevy commented:
@jaymode I have added tests to verify that is_write_index is successfully resolved when aliases point to multiple indices under security.

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -255,6 +271,24 @@ public Index concreteSingleIndex(ClusterState state, IndicesRequest request) {
+     * @return the write index obtained as a result of the index resolution
+     */
+    public Index concreteWriteIndex(ClusterState state, IndicesRequest request) {
+        String indexExpression = request.indices() != null && request.indices().length > 0 ? request.indices()[0] : null;
```
I don't think null is good hear (it translates to an all).

talevy replied:
why would this be different than concreteSingleIndex?

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -255,6 +271,24 @@ public Index concreteSingleIndex(ClusterState state, IndicesRequest request) {
+        Context context = new Context(state, request.indicesOptions(), false, true);
+        Index[] indices = concreteIndices(context, indexExpression);
+        // concreteIndices will throw its own exception when checking for a write index. Assert here for good measure.
+        assert indices.length == 1 : "The index/alias and options provided did not point to a write-index";
```
I think this needs to be stronger - the expression resolvers can in theory give you multiple expressions / indices. This needs to be a hard exception (and we should test it)

talevy replied:
you're right, need more tests for multiple index expressions

talevy replied:
after looking at this for a second... looks like the intent of this method is to be called against DocWriteRequests, which, whether they actually have multiple expressions or not, the above lines filter for the first index expression. So conreteIndices will never be called with multiple index expressions, even if the request did have them. That is why I could never reach this situation. With write index being requested, concreteIndices will throw an exception if nothing is found, otherwise it will add the one write index and continue on

talevy replied:
Even though DocWriteRequest makes sense here, it is not up to the DocWriteRequest to specify that it would

1. have only one index expression in indices()
2. use IndicesOptions.strictSingleIndexNoExpandForbidClosed()

So, leaving it as IndicesRequest is just as good it seems

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -487,32 +488,34 @@ public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex)
         }
         AliasOrIndex.Alias alias = (AliasOrIndex.Alias) result;
+
+        List<String> indexReferencesWithRouting = new ArrayList<>();
```
can we separate this into a different change, so we can discuss it?

talevy replied:
I will try to split it out. I think it may require a few mini changes to make things pass in the meantime.

bleskes commented on server/src/test/java/org/elasticsearch/aliases/IndexAliasesIT.java:
```diff
@@ -98,6 +109,34 @@ public void testAliases() throws Exception {
+        DeleteResponse deleteResponse = client().delete(deleteRequest("alias1").type("type1").id("1")).actionGet();
+        assertThat(deleteResponse.getIndex(), equalTo("test_x"));
+
         logger.info("--> remove [alias1], Aliasing index [test_x] with [alias1]");
```
we now lost the part of the test that goes from two aliases without a flag to one without a flag.

talevy replied:
you're right, re-added

bleskes commented on server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java:
```diff
@@ -996,6 +1002,81 @@ public void testIndexAliases() {
+                .putAlias(AliasMetaData.builder("test-alias").writeIndex(randomFrom(false, null))));
+        ClusterState state = ClusterState.builder(new ClusterName("_name")).metaData(mdBuilder).build();
+        String[] strings = indexNameExpressionResolver
+            .indexAliases(state, "test-0", x -> true, true, "test-*");
```
is have no write flags set on an index tested somewhere here?

talevy replied:
what do you mean by this? no `is_write_index` flag set?

bleskes replied:
got confused. sorry for noise.

talevy commented:
Hi @bleskes. I got back from vacation and picked up with your review.

- I've added an extra test for checking resolution of the write index when moving from two is_write_index:null aliases is reduced to one is_write_index:null
- I've reverted *some* of the routing resolution logic from Metadata. For things to still work properly, a flag is necessary. I know the previous implementation tried to be smart about when it would allow multiple indices to be routed, which was debatably the right thing to do. Now, an `isWriteOperation` flag is passed in so that write operations pointing to multiple indices will simply return the routing value and will not retrieve them from the aliases (since which alias is debatable).
I will open up an issue to discuss this

talevy commented:
Hey @bleskes,

- I've separated out the routing resolution to write resolution vs. read resolution (existing) as we 
discussed off-github. Now routing is validated and resolved against the write index (and it must exist). This means that existing behavior is preserved for when there is an alias pointing to one index (and `is_write_index != false`). 
- I've hardened the checks around writeIndex, requiring that the IndiceRequest have exactly one index expression
- added bulk action tests to double check that they are hooked into the new write-specific routing evaluation

thanks!

bleskes commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -471,6 +471,42 @@ private static String mergePaths(String path, String field) {
+        }
+        AliasMetaData aliasMd = writeIndex.getAliases().get(alias.getAliasName());
+        if (aliasMd.indexRouting() != null) {
+            if (aliasMd.indexRouting().indexOf(',') != -1) {
```
I wonder if we should validate that when people set is_write_index to true as well

talevy replied:
I think that to do this validation justice, it will have to be looked at as something that can be invalid for both read and write. Since this check is similar in both. For this reason, I would feel better if that stricter validation be looked at outside this PR

bleskes replied:
ok. I agree it's not straight forward. Let's have another PR (I tend to say that multiple routing values changes the default for is_write_index and disallows setting it)

bleskes commented on server/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionTests.java:
```diff
@@ -132,4 +160,55 @@ public void testDeleteNonExistingDocExternalGteVersionCreatesIndex() throws Exce
         }));
     }
+
+    public void testRoutingResolvesToWriteIndexAliasMetaData() throws Exception {
```
I wonder if we need a test at this level. We already unit test MetaData and we have an IT test to check it works end to end.

bleskes replied:
actually, I don't think we have an IT test, but we should

talevy replied:
I couldn't find an IT test that verifies this, and I thought that this level of unit test was sufficient. Will add IT test (will also double check that one doesn't exist)

bleskes replied:
I think this can be removed now that we have an IT test? this layer doesn't add anything to the write index resolving as it is all in the resolver? it's just glue and is covered well by the IT?

bleskes commented on server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java:
```diff
@@ -996,6 +1004,149 @@ public void testIndexAliases() {
+        Arrays.sort(strings);
+        assertArrayEquals(new String[] {"test-alias"}, strings);
+        IllegalArgumentException exception = expectThrows(IllegalArgumentException.class,
+            () -> indexNameExpressionResolver.concreteIndexNames(state, IndicesOptions.strictSingleIndexNoExpandForbidClosed(),
```
this doesn't specify anything concerning a write request but you never set the write flag on the aliases. I think we can also randomize setting it?

talevy replied:
you're right. this test was not added for checking write index resolution. This test was added because there was no unit test that hit this branch of the code beforehand. I will add a write-index to the mix for completion

bleskes commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -172,6 +172,76 @@ public void testResolveIndexRouting() {
+            .numberOfShards(1)
+            .numberOfReplicas(0)
+            .putAlias(AliasMetaData.builder("alias0").build())
+            .putAlias(AliasMetaData.builder("alias1").routing("1").build())
```
we don't have an explicit is_write_index set to true in this test. Set it here randomly?

talevy replied:
didn't hurt to add. added

bleskes replied:
I think it's good to also have `null` (and your change sometimes sets it to `false` which isn't identical to what it did before, but I presume that's not a problem)

bleskes commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java:
```diff
@@ -172,6 +172,76 @@ public void testResolveIndexRouting() {
+            .numberOfReplicas(0)
+            .putAlias(AliasMetaData.builder("alias0").build())
+            .putAlias(AliasMetaData.builder("alias1").routing("1").build())
+            .putAlias(AliasMetaData.builder("alias2").routing("1,2").build())
```
I don't believe we check rejection of multi value routings with is_write_index set to true. Can we add a test that it fails (preferably but rejecting it when building the metadata objects?)

talevy replied:
I will add a new `alias4` with `is_write_index=true` and have it fail.

> (preferably but rejecting it when building the metadata objects?)

I think this touches the previous comment about failing invalid routing values early. Although I agree, I do not feel comfortable introducing this change in this PR

bleskes replied:
++

talevy commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -193,30 +192,44 @@ public IndexNameExpressionResolver(Settings settings) {
+                AliasOrIndex.Alias alias = (AliasOrIndex.Alias) aliasOrIndex;
+                IndexMetaData writeIndex = alias.getWriteIndex();
+                if (writeIndex == null) {
+                    if (alias.getIndices().size() > 1) {
```
I will make more generic

talevy replied:
> it may be that all of them have their write index flag set to false

saying it points to multiple indices with none set as a write-index [is_write_index=true] does not conflict with that statement?

bleskes replied:
isn't this covered by "the alias points to multiple indices without one being designated as a write index"?

bleskes commented on server/src/test/java/org/elasticsearch/action/bulk/BulkIntegrationIT.java:
```diff
@@ -40,4 +48,51 @@ public void testBulkIndexCreatesMapping() throws Exception {
+        assertThat(getBuilder.get().getSource().get("foo"), equalTo("bar"));
+        assertThat(client().prepareGet("index3", "type", "id").setRouting("1").get().getSource().get("foo"), equalTo("baz"));
+
+        client().prepareBulk().add(client().prepareUpdate("alias1", "type", "id").setDoc("foo", "updated")).get();
```
we should check errors here?

talevy replied:
will add. didn't think this was necessary since there are get-action calls verifying success of this operation

bleskes commented on server/src/test/java/org/elasticsearch/action/bulk/BulkIntegrationIT.java:
```diff
@@ -40,4 +48,51 @@ public void testBulkIndexCreatesMapping() throws Exception {
+        client().prepareBulk().add(client().prepareUpdate("alias1", "type", "id").setDoc("foo", "updated")).get();
+        assertThat(getBuilder.get().getSource().get("foo"), equalTo("bar"));
+        assertThat(client().prepareGet("index3", "type", "id").setRouting("1").get().getSource().get("foo"), equalTo("updated"));
+        client().prepareBulk().add(client().prepareDelete("alias1", "type", "id")).get();
```
I'm not sure what we're testing here?

talevy replied:
TransportBulkAction resolves routing for delete requests just as it does for index and update. This delete request is here to verify this

bleskes commented on server/src/test/java/org/elasticsearch/action/bulk/BulkIntegrationIT.java:
```diff
@@ -40,4 +48,51 @@ public void testBulkIndexCreatesMapping() throws Exception {
+            indexRequest.routing("0");
+        }
+        indexRequest.source(Collections.singletonMap("foo", "bar"));
+        IndexRequest indexRequestWithAlias = new IndexRequest("alias1", "type", "id");
```
can we use different ids for the different indices? I find this super confusing to reason about. Maybe also add the routing value you expect to be used to the id.

talevy replied:
I think it is also OK to remove the first index request entirely. I just wanted to it to demonstrate that the two calls do not interact with the same document, even though they have the same index/type/id.

talevy replied:
On second thought, I remember the purpose of this other index request. It is to verify the other branches of the routing resolution. The reason I think it is OK to remove is that this IT test is meant to test the connection to this method, not all its branches. The Unit tests do that.

bleskes commented on server/src/test/java/org/elasticsearch/action/bulk/BulkIntegrationIT.java:
```diff
@@ -40,4 +47,35 @@ public void testBulkIndexCreatesMapping() throws Exception {
+        client().admin().indices().prepareCreate("index3")
+            .addAlias(new Alias("alias1").indexRouting("1").writeIndex(true)).setSettings(twoShardsSettings).get();
+
+        IndexRequest indexRequestWithAlias = new IndexRequest("alias1", "type", "id");
```
randomly set the routing to "1"?

talevy commented:
thanks for the thorough review iterations @bleskes
